quantization/nlp/bert/trt/README.md

```diff
@@ -0,0 +1,63 @@
+* **extra_options** : Default is {}. It can accept `num_bins`, `num_quantized_bins` and `percentile` as options. If no options are given, it will use internal default settings to run the calibration. When using entropy algorithm, `num_bins` (means number of bins of histogram for collecting floating tensor data) and `num_quantized_bins` (means number of bins of histogram after quantization) can be set with different combinations to test and fine-tune the calibration to get optimal result, for example, {'num_bins':8001, 'num_quantized_bins':255}. When using percentile algorithm, `num_bins` and `percentile` can be set with different values to fine-tune the calibration to get better result, for example, {'num_bins':2048, 'percentile':99.999}. 
+
+## QDQ Model Generation
+In order to get best performance from TensorRT, there are some optimizations being done when inserting Q/DQ nodes to the model.
+* When inserting QDQ nodes to 'Add' node, only insert QDQ nodes to 'Add' node which is followed by ReduceMean node
+* Enable per channel quantization on 'MatMul' node's weights. Please see `QDQQuantizer(...)` in the e2e example, the per_channel argument should be True as well as 'QDQOpTypePerChannelSupportToAxis': {'MatMul': 1} should be specified in extra_options argument. You can also modify 'QDQOpTypePerChannelSupportToAxis' to other op types and channel axis if they can increase performance.
+
+Once QDQ model generation is done, the qdq_model.onnx will be saved.
+
+## QDQ Model Evaluation
+Remember to set env variables, ORT_TENSORRT_FP16_ENABLE=1 and ORT_TENSORRT_INT8_ENABLE=1, to run QDQ model.
+We use evaluation tool from Nvidia TensorRT demo BERT repo to evaluate the result based on SQuAD v1.0 and SQuAD v2.0.
+
+Note: The input names of model in the e2e example is based on Hugging Face Model's naming. If model input names are not correct in your model, please modify the code ort_session.run(["output_start_logits","output_end_logits"], inputs) in the example.
+
+## Performance
+The performance results were obtained by running [onnxruntime_perf_test](https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/test/perftest) with TensorRT 8.2.2.1 on
+NVIDIA T4 with (1x T4 32G) GPUs. 
+```shell
+./onnxruntime_perf_test -e tensorrt -r 10000 qdq_model.onnx -o 0 -i 'trt_fp16_enable|true trt_int8_enable|true trt_engine_cache_enable|true'
+```
+
+
+The accuracy results were obtained by running e2e_tensorrt_bert_example.py as above mentioned. 
+#### BERT Base
+| Sequence Length | Batch Size | FP32 Latency (ms) | FP16 Latency (ms) |  INT8 Latency (ms) | FP32 Accuracy (F1) | FP16 Accuracy (F1) |    INT8 Accuracy (F1) |
+|-----------------|------------|----|-------------|--------|------|-------|---------|
+| 128 | 1 | 7.33134 ms| 2.14245 ms | 1.65864 ms |88.259| 88.111 | 82.988 |
+| 384 | 1 | 21.6183 ms| 4.55848 ms | 3.58168 ms |88.662| 88.595 | 82.739 |
```

Comment:
It seems strange that accuracy of sequence length 384 is less than 128.

Reply:
I generated the two qdq models with sequence length 384 and sequence length 128 again. Still got the same accuracy numbers. However, when I increased the calibration number from 100 to 200, the accuracy number of sequence length 384 increases to almost 82.988. So I'm thinking with larger calibration number, sequence length 384 could have better accuracy than sequence length 128.